drivers/i2c/i2c_sam0.c

```diff
@@ -155,7 +155,20 @@ static void i2c_sam0_isr(const struct device *dev)
 	if (status & SERCOM_I2CM_INTFLAG_MB) {
 		if (!data->msg.size) {
 			i2c->INTENCLR.reg = SERCOM_I2CM_INTENCLR_MASK;
-			i2c->CTRLB.bit.CMD = 3;
+
+			/*
+			 * writing to i2c->ADDR for the next message will cause a repeated
+			 * start condition. this is different to continue_next, as the
+			 * data->msg.size is already at zero (not one), and
+			 * i2c_sam0_transfer() is responsible for advancing to the next
+			 * message, not this ISR.
+			 */
+			if ((data->num_msgs <= 1)
+			    || (data->msgs[0].flags & I2C_MSG_STOP)
+			    || !(data->msgs[1].flags & I2C_MSG_RESTART)) {
+				i2c->CTRLB.bit.CMD = 3;
```

Comment:
Since this is a complex operation I don't mind to have a description about what 3 means, like

```diff
+				/* 3 - Execute acknowledge action succeeded
+				 *     by issuing a Stop condition
+				 */
				i2c->CTRLB.bit.CMD = 3;
```